File: app/src/main/java/com/mecong/tenderalarm/sleep_assistant/SleepAssistantPlayListModel.java
package com.mecong.tenderalarm.sleep_assistant;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.mecong.tenderalarm.sleep_assistant.media_selection.SleepMediaType;

import java.util.Collections;
import java.util.List;


public class SleepAssistantPlayListModel extends ViewModel {
    MutableLiveData<PlayList> playlist = new MutableLiveData<>();
    MutableLiveData<Boolean> playing = new MutableLiveData<>();

    LiveData<PlayList> getPlaylist() {
        return playlist;
    }

    public void setPlaylist(PlayList playlist) {
        this.playlist.setValue(playlist);
    }

    public SleepAssistantPlayListModel(MutableLiveData<PlayList> playlist, MutableLiveData<Boolean> playing) {
        this.playlist = playlist;
        this.playing = playing;
    }

    public SleepAssistantPlayListModel() {
    }

    //}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}]


    public static class PlayList {
        int index;
        List<Media> media;
        SleepMediaType mediaType;

        public PlayList(String url, String name, SleepMediaType mediaType) {
            this.media = Collections.singletonList(new Media(url, name));
            this.mediaType = mediaType;
            index = 0;
        }

        public PlayList(int index, List<Media> media, SleepMediaType mediaType) {
            this.index = index;
            this.media = media;
            this.mediaType = mediaType;
        }

        public int getIndex() {
            return index;
        }

        public List<Media> getMedia() {
            return media;
        }

        public SleepMediaType getMediaType() {
            return mediaType;
        }
    }


    public static class Media {
        private String url;
        private String title;

        public Media(String url, String title) {
            this.url = url;
            this.title = title;
        }

        public String getUrl() {
            return url;
        }

        public String getTitle() {
            return title;
        }
    }
}

File: app/src/main/java/com/mecong/tenderalarm/sleep_assistant/SleepAssistantPlayListModel.kt
package com.mecong.tenderalarm.sleep_assistant

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.mecong.tenderalarm.sleep_assistant.media_selection.SleepMediaType

class SleepAssistantPlayListModel : ViewModel() {
  var playlist = MutableLiveData<SleepAssistantPlayList>()

  var playing = MutableLiveData<Boolean>()
}


data class Media(val url: String, val title: String)

class SleepAssistantPlayListActive(index: Int, media: List<Media>, mediaType: SleepMediaType, playListId: Long) :
  SleepAssistantPlayList(index, media, mediaType, playListId)

class SleepAssistantPlayListIdle(index: Int, media: List<Media>, mediaType: SleepMediaType, playListId: Long) :
  SleepAssistantPlayList(index, media, mediaType, playListId)

open class SleepAssistantPlayList(
  var index: Int,
  val media: List<Media>,
  val mediaType: SleepMediaType,
  val playListId: Long
) {

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false
    other as SleepAssistantPlayList
    if (index != other.index) return false
    if (mediaType != other.mediaType) return false
    if (playListId != other.playListId) return false
    return true
  }

  override fun hashCode(): Int {
    var result = index
    result = 31 * result + media.hashCode()
    result = 31 * result + mediaType.hashCode()
    result = 31 * result + playListId.hashCode()
    return result
  }
}
